### PR TITLE
usb: cdc_acm: Additional connection state checks

### LINKED ISSUE #72786: cdc acm drops packets silently
**Describe the bug**
When sending usb packets continuously apparently some are dropped.
There are no logs from the usb subsystem that bytes were dropped.
I was testing on a nrf52840dk with the cdc_acm_echo sample.

**To Reproduce**
1. edit build the cdc_acm sample
2. send chunks of increasing bytes (e.g 128 bytes) and read them back in a loop
3. observe missing bytes in the sequence after some time

Adapted interrupt handler
```
uint8_t tx_byte = 0x00;
bool rx_err = false;
bool tx_err = false;

static void interrupt_handler(const struct device *dev, void *user_data)
{
	ARG_UNUSED(user_data);

	while (uart_irq_update(dev) && uart_irq_is_pending(dev)) {
		if (!rx_throttled && uart_irq_rx_ready(dev)) {
			int recv_len, rb_len;
			uint8_t buffer[64];
			size_t len = MIN(ring_buf_space_get(&ringbuf),
					 sizeof(buffer));

			if (len == 0) {
				/* Throttle because ring buffer is full */
				uart_irq_rx_disable(dev);
				rx_throttled = true;
				continue;
			}

			recv_len = uart_fifo_read(dev, buffer, len);
			if (recv_len < 0) {
				LOG_ERR("Failed to read UART FIFO");
				recv_len = 0;
			};

			for (int i = 0; i < recv_len; i++) {
				if (!rx_err && (buffer[i] != rx_byte)) {
					LOG_ERR("Received byte does not match expected byte! %02X != %02X", buffer[i], rx_byte);
					rx_err = true;
				}
				rx_byte++;
			}

			rb_len = ring_buf_put(&ringbuf, buffer, recv_len);
			if (rb_len < recv_len) {
				LOG_ERR("Drop %u bytes", recv_len - rb_len);
			}

			LOG_DBG("tty fifo -> ringbuf %d bytes", rb_len);
			if (rb_len) {
				uart_irq_tx_enable(dev);
			}
		}

		if (uart_irq_tx_ready(dev)) {
			uint8_t buffer[64];
			int rb_len, send_len;

			rb_len = ring_buf_get(&ringbuf, buffer, sizeof(buffer));
			if (!rb_len) {
				LOG_DBG("Ring buffer empty, disable TX IRQ");
				uart_irq_tx_disable(dev);
				continue;
			}

			for (int i = 0; i < rb_len; i++) {
				if (!tx_err && (buffer[i] != tx_byte)) {
					LOG_ERR("Sent byte does not match expected byte! %02X != %02X", buffer[i], tx_byte);
					tx_err = true;
				}
				tx_byte++;
			}

			if (rx_throttled) {
				uart_irq_rx_enable(dev);
				rx_throttled = false;
			}

			send_len = uart_fifo_fill(dev, buffer, rb_len);
			if (send_len < rb_len) {
				LOG_ERR("Drop %d bytes", rb_len - send_len);
			}

			LOG_DBG("ringbuf -> tty fifo %d bytes", send_len);
		}
	}
}```

**Expected behavior**
No bytes being dropped

**Impact**
Showstopper for a usb to serial converter application

**Logs and console output**
I added byte counters to the sample to observe on which side the bytes are being dropped.
Apparently they are dropped on the usb receive side.

```
*** Booting Zephyr OS build v3.6.0-3610-g5cbf7b1b9d13 ***
[00:00:00.250,183] <inf> cdc_acm_echo: Wait for DTR
[00:00:00.254,211] <inf> usb_cdc_acm: Device suspended
[00:00:00.541,503] <inf> usb_cdc_acm: Device resumed
[00:00:00.727,050] <inf> usb_cdc_acm: Device configured
[00:00:02.751,922] <inf> cdc_acm_echo: DTR set
[00:00:02.869,049] <inf> cdc_acm_echo: Baudrate 115200
[00:02:35.662,628] <err> cdc_acm_echo: Received byte does not match expected byte! C0 != 80
[00:02:35.662,689] <err> cdc_acm_echo: Sent byte does not match expected byte! C0 != 80
```

Also the logs from the test script:

```
Starting echo test on /dev/ttyACM2 (3)
baud=115200, size=128
Received byte does not match expected byte: C0 != 80
Last sent on 3:
80 81 82 83 84 85 86 87 88 89 8A 8B 8C 8D 8E 8F 90 91 92 93 94 95 96 97 98 99 9A 9B 9C 9D 9E 9F
A0 A1 A2 A3 A4 A5 A6 A7 A8 A9 AA AB AC AD AE AF B0 B1 B2 B3 B4 B5 B6 B7 B8 B9 BA BB BC BD BE BF
C0 C1 C2 C3 C4 C5 C6 C7 C8 C9 CA CB CC CD CE CF D0 D1 D2 D3 D4 D5 D6 D7 D8 D9 DA DB DC DD DE DF
E0 E1 E2 E3 E4 E5 E6 E7 E8 E9 EA EB EC ED EE EF F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 FA FB FC FD FE FF

Last received on 3:
C0 C1 C2 C3 C4 C5 C6 C7 C8 C9 CA CB CC CD CE CF D0 D1 D2 D3 D4 D5 D6 D7 D8 D9 DA DB DC DD DE DF
E0 E1 E2 E3 E4 E5 E6 E7 E8 E9 EA EB EC ED EE EF F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 FA FB FC FD FE
Echo test terminated after 153 seconds
```

It seems like the first 64 bytes are dropped on the usb receive side of the nrf52.

**Environment (please complete the following information):**
 - OS of the test host: Debian 11 (bullseye)
 - Zephyr SDK; 0.16.3
 - Zephyr commit: 5cbf7b1b9d13a39f36ad8dabfaeaec21d98cf5ff

### PR DESCRIPTION
**Pull request by metratec-mkoehler**
Issues caused:
- On intense use I had packet losses (below 1 per 1000)
- Restarting the host device or resetting the USB controller caused the usb transfer to get stuck in about 30% of tests
- This includes issue #72786 opened by a collegue

Analysis of issues:
  I found multiple cases causing this isssues with wrong return values on cdc_acm_irq_is_pending(), cdc_acm_irq_tx_ready() etc causing my logic to get stuck

Changes:
- usb_transfer(): k_work_submit_to_queue() return value was ignored -> calling function tx_work_handler() dropped data even though it was
   not put to workqueue
- tx_work_handler() put dev_data->tx_ready to false even if usb_transfer() failed but without the transfer done callback it was never reset to true so TX got stuck
- cdc_acm_irq_tx_ready() did not check configured and suspended states This caused the device to show as ready even though it was not configured after a host reset
- cdc_acm_irq_rx_ready() did not check if irq is enabled
- cdc_acm_irq_is_pending() had the same issues as its just a test on cdc_acm_irq_rx_ready() or cdc_acm_irq_tx_ready() (the description says so, too) -> Replaced it by a call of both functions for a fast and consistent fix

Tests:
- The host reset issue was tested with a Intel NUC with Windows 11 in an endless reset loop, about 2 times per minute, for 12 hours. The issue no longer applies there
- The host reset issue was also tested on a Linux XUbuntu host with a pyusb device reset script. The issue was tested only manually (at 50% success that was good enough) and does not apply any longer
- Data loss was tested with an echo device and a data generating and checking script. It was also tested with a device generating serialized data (counting up).
- Both showed data loss (about 1 byte per GB, in blocks of multiples of 8). For both test cases the issue no longer applied (tested with more than 1 TB of data) Fixes: #72786

Considerations:
- Changes in cdc_acm.c should not have impact on other code. Return values may show an error that was missed before
- Change in usb_transfer.c may influence more devices The new return value should be in an expected range (0 or negative) so I expect it to work with all other components as well

### CODE REVIEW

**Comment by metratec-mkoehler** [comment]
Failing due to issue https://github.com/zephyrproject-rtos/zephyr/issues/73846

**Comment by metratec-mkoehler** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/73838

Needed

File: subsys/usb/device/usb_transfer.c [hunk 1]
```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
 
 	if (flags & USB_TRANS_WRITE) {
 		/* start writing first chunk */
-		k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
+		ret = k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
+		ret = MAX(0, ret);
```
Comment (josuah) [inline subsys/usb/device/usb_transfer.c:255]: I often mix min() an max(): `MAX(0, ret)` would take the greatest value between `0` and `ret`, so if `ret` is negative (i.e. error), and only pick 0, 1 or 2 as described here:
https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#c.k_work_submit_to_queue

Is that what is intended?
Thanks

Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:255]: Thanks for the feedback. 
You are totally right, this is an error. Unsurprisingly the work queue never caused an issue in tests.
Fix incoming.

File: subsys/usb/device/usb_transfer.c [hunk 2]
```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
 
 	if (flags & USB_TRANS_WRITE) {
 		/* start writing first chunk */
-		k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
+		ret = k_work_submit_to_queue(&USB_WORK_Q, &trans->work);
```
Comment (tmon-nordic) [inline subsys/usb/device/usb_transfer.c:254]: Please split this to separate commit with its own commit message explaining what exactly it does.

**Review by tmon-nordic — CHANGES_REQUESTED** [review]
Looks good. Please separate the CDC ACM changes and generic transfer handling changes - while together these do solve the same observed issue, they are really different root causes leading to the same symphtoms.

File: subsys/usb/device/usb_transfer.c (hunk 2 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:254]: Done

**Comment by tmon-nordic** [comment]
You need to reword the commit messages. For instance the
```
- Change in usb_transfer.c may influence more devices
  The new return value should be in an expected range (0 or negative)
  so I expect it to work with all other components as well
```
is no longer relevant in the first commit.

The commit message format looks like originating from very formal process.

The second commit fails to really state what the change achieves and is mostly a copy of the first commit.

**Comment by metratec-mkoehler** [comment]
The commit message format originates from
https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribute-guidelines
> 
    what the change does,
    why you chose that approach,
    what assumptions were made, and
    how you know it works – for example, which tests you ran.

But I will save me time in the future keeping things minimal instead. Change will come soon.

**Review by tmon-nordic — DISMISSED** [review]

File: subsys/usb/device/class/cdc_acm.c [hunk 3]
```diff
@@ -288,10 +287,13 @@ static void tx_work_handler(struct k_work *work)
...
-		     cdc_acm_write_cb, dev_data);
-
-	ring_buf_get_finish(dev_data->tx_ringbuf, len);
+	ret = usb_transfer(ep, data, len, USB_TRANS_WRITE, cdc_acm_write_cb, dev_data);
+	if (ret < 0) {
+		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
+		return;
+	}
+	dev_data->tx_ready = false;
+	(void)ring_buf_get_finish(dev_data->tx_ringbuf, len);
```
Comment (jfischer-no) [inline subsys/usb/device/class/cdc_acm.c:296]:
```c
	ret = usb_transfer(ep, data, len, USB_TRANS_WRITE,
			   cdc_acm_write_cb, dev_data);
	if (ret != 0) {
		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
		return;
	}

	dev_data->tx_ready = false;
	(void)ring_buf_get_finish(dev_data->tx_ringbuf, len);
```

File: subsys/usb/device/class/cdc_acm.c [hunk 4]
```diff
@@ -621,13 +623,9 @@ static void cdc_acm_irq_tx_disable(const struct device *dev)
...
 {
-	struct cdc_acm_dev_data_t * const dev_data = dev->data;
+	struct cdc_acm_dev_data_t *const data = dev->data;
 
-	if (dev_data->tx_irq_ena && dev_data->tx_ready) {
-		return 1;
-	}
-
-	return 0;
+	return data->tx_irq_ena && data->configured && data->tx_ready && !data->suspended;
```
Comment (jfischer-no) [inline subsys/usb/device/class/cdc_acm.c:628]: I do not think it is right, whether configured or not, UART device itself can fill data into FIFO, it will also break established expectations.

File: subsys/usb/device/usb_transfer.c [hunk 5]
```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
```
Comment (jfischer-no) [inline subsys/usb/device/usb_transfer.c:251]: Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.

>Sometimes loosing USB transfers

Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?

>Tests with echo device

What is "echo device"?

There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/usb/device/class/cdc_acm.c (hunk 3 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:296]: The line break is according to [the new formatting rule](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style)  (100 chars per line).

Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:296]: Also, ret < 0 is more robust to API changes than ret != 0 as giving back the number of finished bytes might be a future way to implement. If you want a security against an unexpected positive return value I would be fine with adding an
```
 __ASSERT(ret <= 0, "Unexpected return value");
```

File: subsys/usb/device/class/cdc_acm.c [hunk 6]
```diff
@@ -256,6 +256,7 @@ static void tx_work_handler(struct k_work *work)
 	uint8_t ep = cfg->endpoint[ACM_IN_EP_IDX].ep_addr;
```
[the remark this answers is not in the document]
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:256]: As I say in the initial commit (which I was forced to split and shorten):

usb_transfer() drops data by not checking the return value of k_work_submit_to_queue()
This causes tx_work_handler() to not send the data because it thinks they are in the work queue now. It also thinks the device is occupied but as there is no work queue item it won't get freed again. 

The stuck after reset comes from:
cdc_acm_irq_tx_ready() did not check configured and suspended states 
-> This caused the device to show as ready even though it was not configured after a host reset
The same can happen for RX so I fixed it with it, it's the same issue of ignored state.

Together these cause the issues described in the issue post. 
My changes changed this.

File: subsys/usb/device/class/cdc_acm.c (hunk 4 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:628]: I think it is right.

cdc_acm_fifo_fill() will not accept data. 

```
	if (!dev_data->configured || dev_data->suspended) {
		LOG_WRN("Device not configured or suspended, drop %d bytes",
			len);
		return 0;
	}
```

This cause an inifinte loop when using cdc_acm_irq_tx_ready() as in
zephyr/samples/subsys/usb/cdc_acm/src/main.c interrupt_handler() and many other projects.

File: subsys/usb/device/usb_transfer.c (hunk 5 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:251]: > Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.

But the 2 commits (formerly one) DO fix the issue. But maybe only adding it to the second commit is fine, too.

Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:251]: > Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?

usb_transfer() returns zero (success) even though the queueing failed for one of various reasons. This causes tx_work_handler() to silently drop data (which is the issue described in #72786)

Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:251]: > What is "echo device"?

A board with samples/subsys/usb/cdc_acm firmware (echoing data to the same USB CDC ACM port)

Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:251]: > There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.

I really don't care so much if you doubt it. The data gets lost, I can see it on wireshark, I can find the reason with a debugger, I can reproduce it (and no, not in 5 minutes, as it's random and was about 1 byte / GB in my case) and the changes fix it. 

Also the description shows the use of functions was just wrong (ignoring return values and states) and should be fixed anyway. With my issue or without it. I am happy to improve my commits as needed and did already add quite some work to this fix but it's time to finish.

File: subsys/usb/device/class/cdc_acm.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/usb/device/class/cdc_acm.c:256]: > As I say in the initial commit (which I was forced to split and shorten):
> 
> usb_transfer() drops data by not checking the return value of k_work_submit_to_queue() This causes tx_work_handler() to not send the data because it thinks they are in the work queue now. It also thinks the device is occupied but as there is no work queue item it won't get freed again.

Can you please provide k_work_submit_to_queue() return value instead of your own interpretation?

> The stuck after reset comes from: cdc_acm_irq_tx_ready() did not check configured and suspended states -> This caused the device to show as ready even though it was not configured after a host reset The same can happen for RX so I fixed it with it, it's the same issue of ignored state.

But it is okay to fill/read the UART FIFO (ringbuffer) event if USB device is not configured or suspended.

> 
> Together these cause the issues described in the issue post. My changes changed this.

I can help you write a commit message later, after we figure out what caused the issue.

[the remark this answers is not in the document]
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:256]: > I can help you write a commit message later, after we figure out what caused the issue.

That's great, learning to fit better into the zephyr workflow will be helpful.  :)
It may delay because today is my last day before the holidays but with some luck it may be done before

File: subsys/usb/device/class/cdc_acm.c (hunk 3 above)
Comment (jfischer-no) [reply subsys/usb/device/class/cdc_acm.c:296]: > The line break is according to [the new formatting rule](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style) (100 chars per line).

Sure ("The line length is 100 columns or fewer"), but I am asking you to change it in a minimal way without unnecessary reformatting.



> Also, ret < 0 is more robust to API changes than ret != 0 as giving back the number of finished bytes might be a future way to implement. If you want a security against an unexpected positive return value I would be fine with adding an
> 
> ```
>  __ASSERT(ret <= 0, "Unexpected return value");
> ```

No, there is no future for `usb_transfer()`, and I am asking for exactly that
```c
	if (ret != 0) {
		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
		return;
	}
```

File: subsys/usb/device/usb_transfer.c (hunk 5 above)
Comment (jfischer-no) [reply subsys/usb/device/usb_transfer.c:251]: > > Please remove "Partially fixes issue #72786" from the commit message.
> 
> But the 2 commits (formerly one) DO fix the issue. But maybe only adding it to the second commit is fine, too.

You can add Fixes #72786 to PR description, but "Partially fixes issue #72786" should be removed from commit message.

> > Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?
> 
> usb_transfer() returns zero (success) even though the queueing failed for one of various reasons. This causes tx_work_handler() to silently drop data (which is the issue described in #72786)

You still do not provide a description for why k_work_submit_to_queue() failed.

> > There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.
> 
> I really don't care so much if you doubt it. The data gets lost, I can see it on wireshark, I can find the reason with a debugger, I can reproduce it (and no, not in 5 minutes, as it's random and was about 1 byte / GB in my case) and the changes fix it.

But you should care. Exact steps and environment to reproduce it is still missing tought it was requested few times. The reason why k_work_submit_to_queue() sometimes fails is unknown.

File: subsys/usb/device/class/cdc_acm.c (hunk 3 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:296]: I will change it

File: subsys/usb/device/class/cdc_acm.c (hunk 4 above)
Comment (jfischer-no) [reply subsys/usb/device/class/cdc_acm.c:628]: cdc_acm_irq_tx_ready() is implementation of uart_irq_tx_ready(), where the descriptions is "Check if UART TX buffer can accept a new char", `TX buffer`, it is not `Check if UART controller can transfer a new char`.

File: subsys/usb/device/class/cdc_acm.c (hunk 3 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c]: Changed. Also reordered the commits as the fix on cdc acm depends on the usb_transfer() fix so first transfer, then the cdc acm. The second one now states correctly that it fixes the issue.

File: subsys/usb/device/class/cdc_acm.c (hunk 4 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/class/cdc_acm.c:630]: Well... but it can't (or at least: won't). It will always return 0 because it will always take zero chars. That's not accepting a new char.

File: subsys/usb/device/usb_transfer.c (hunk 5 above)
Comment (metratec-mkoehler) [reply subsys/usb/device/usb_transfer.c:251]: > You still do not provide a description for why k_work_submit_to_queue() failed.

Ah, I didn't get this was even the question, I thought it was about what happens after the error response, not before. My bad.

 I don't know what the error code is but it's very likely -EBUSY (as the other ones are for uninit device, and it's working for the overwhelming number of packets). And I don't see why it is relevant to the fix. On the programmer side I understand the need to go deeper into the root cause, but getting busy response from the queue doesn't seem to be something to bother at a rate of less than 1 per billion tries. Is it that unexpected?

At this low error rate finding the issue would take days or weeks without any real win. There can be a new issue to investigate, but for the USB data loss this is the fix and it took already quite some time to find the issues.

File: subsys/usb/device/class/cdc_acm.c (hunk 4 above)
Comment (tmon-nordic) [reply subsys/usb/device/class/cdc_acm.c:630]: The suspend logic in CDC ACM class implementation itself is not quite right.

Linux will suspend the ttyACM device if device claims to support remote wakeup and there is no data from host to device for some time. The device will be woken up by the host if and only if the host wants to send something. If the device has something to send to host, it is up to device to issue to remote wakeup - and it is what Linux expects the ttyACM device supporting remote wakeup will do.

**Comment by metratec-mkoehler** [comment]
I updated the branch according to requests, restructured the commits and pushed that. 
I am back from my vacation and would like to finalize the pull request.

> Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.
Is done by replacing it with "Fixes issue #72786" in the second commit (which does fix it but still depends on the fixed usb_transfer() return value from the first commit)

I will also split the connection check changes from the data loss fix into different commits. 

I would like them to stay in one pull request for convenience and because they got tested together (and I just won't have the time for another test line as the product with the fixed usb code got already shipped), but I can understand if that's not possible. If needed I would split another branch and seperate them from each other. As long as both get pulled this should make no difference.

**Review by tmon-nordic — APPROVED** [review]

File: subsys/usb/device/usb_transfer.c (hunk 5 above)
Comment (jfischer-no) [reply subsys/usb/device/usb_transfer.c:251]: > I don't know what the error code is but it's very likely -EBUSY (as the other ones are for uninit device, and it's working for the overwhelming number of packets). And I don't see why it is relevant to the fix. On the programmer side I understand the need to go deeper into the root cause, but getting busy response from the queue doesn't seem to be something to bother at a rate of less than 1 per billion tries. 

It is important to find the reason, I still doubt that these changes fix your problem.[Here](https://github.com/zephyrproject-rtos/zephyr/pull/73828#discussion_r1645546441
) you wrote that you can reliably reproduce it. Why is it difficult now?

>Is it that unexpected?

Yes, it is unexpected that k_work_submit_to_queue() fails. Btw, what work_queue is used in your configuration?
 
> At this low error rate finding the issue would take days or weeks without any real win. There can be a new issue to investigate, but for the USB data loss this is the fix and it took already quite some time to find the issues.

This PR cannot be considered to fix #72786 until the reason why k_work_submit failed is clear.

File: subsys/usb/device/class/cdc_acm.c (hunk 4 above)
Comment (jfischer-no) [reply subsys/usb/device/class/cdc_acm.c:630]: https://github.com/zephyrproject-rtos/zephyr/pull/73828/commits/490bd42693d5fdd2fa7bfd7c00c34fa076b05e4d#r1646202774 not addressed.

File: subsys/usb/device/class/cdc_acm.c [hunk 7]
```diff
@@ -288,10 +287,15 @@ static void tx_work_handler(struct k_work *work)
...
 	LOG_DBG("Got %zd bytes from ringbuffer send to ep %x", len, ep);
 
-	usb_transfer(ep, data, len, USB_TRANS_WRITE,
-		     cdc_acm_write_cb, dev_data);
+	ret = usb_transfer(ep, data, len, USB_TRANS_WRITE,
+			   cdc_acm_write_cb, dev_data);
+	if (ret != 0) {
+		(void)ring_buf_get_finish(dev_data->tx_ringbuf, 0);
+		return;
+	}
```
Comment (jfischer-no) [inline subsys/usb/device/class/cdc_acm.c:295]: If usb_transfer() fails, it disables TX interrupts until bus reset or similar. Instead, error branch should reschedule tx_work with some delay, maybe 1ms.

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Previous comments are still not addressed.

**Comment by carlescufi** [comment]
@metratec-mkoehler could you please go over @jfischer-no's review comments and address them? Then we may have a chance to get this merged in time for 3.7.0.

**Comment by metratec-mkoehler** [comment]
I talked to my manager and got a go to prepare a test case to support this pull request. 
I will start now and hopefully it's possible to reproduce on my nrf dev board.
I will show the test projects and results asap-

**Comment by metratec-mkoehler** [comment]
I have big issues to reproduce the data loss while logging or using other tools. And it is already annoyingly rare. So I decided to split the reset issue from this pull request. I will make an issue today or tommorow, it's easy to reproduce and the fix should be working without the other changes. 

This way we can have at least one fix in (and it's easier to talk about errors at this point because I have no expertise on the USB protocol itself).

I will add more info on the data loss issue as soon as possible, too.

**Comment by mmahadevan108** [comment]
@metratec-mkoehler, do you have any updates for this PR.

**Comment by metratec-mkoehler** [comment]
Sadly I don't. 

The changes were partially moved into different pull requests 
[https://github.com/zephyrproject-rtos/zephyr/pull/76171] (Host reset)
[https://github.com/zephyrproject-rtos/zephyr/pull/76173] (Logging error, already merged)

I can't reproduce the data loss issue. It is possible the loss was not in Zephyr or, if it was, the changes made in this PR only randomly (race condition?) fixed the issue for me. Other changes, not related to the driver, also "fixed" them in my case. This is highly dissatisfying but it's what it is. 

I will mark this PR as closed (the host reset fix is still pending in the other PR)
